components/brave_new_tab_ui/reducers/rewards_reducer.ts

```diff
@@ -0,0 +1,189 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+import { Reducer } from 'redux'
+
+// Constants
+import { types } from '../constants/new_tab_types'
+
+// API
+import { InitialRewardsData, PreInitialRewardsData } from '../api/initialData'
+import * as storage from '../storage'
+import { getTotalContributions } from '../rewards-utils'
+
+const initialState = storage.load()
+
+export const rewardsReducer: Reducer<NewTab.State | undefined> = (state: NewTab.State | undefined, action: any) => {
+  if (state === undefined) {
+    state = initialState
+  }
+
+  const startingState = state
+  const payload = action.payload
+  switch (action.type) {
+    case types.CREATE_WALLET:
+      chrome.braveRewards.createWallet()
+      state = { ...state }
+      state.rewardsState.walletCreating = true
+      break
+
+    case types.ON_ENABLED_MAIN:
+      state = { ...state }
+      state.rewardsState.enabledMain = payload.enabledMain
+      break
+
+    case types.ON_WALLET_INITIALIZED: {
+      const result: NewTab.RewardsResult = payload.result
+      state = { ...state }
+
+      switch (result) {
+        case NewTab.RewardsResult.WALLET_CORRUPT:
+          state.rewardsState.walletCorrupted = true
+          break
+        case NewTab.RewardsResult.WALLET_CREATED:
+          state.rewardsState.walletCreated = true
+          state.rewardsState.walletCreateFailed = false
+          state.rewardsState.walletCreating = false
+          state.rewardsState.walletCorrupted = false
+          chrome.braveRewards.saveAdsSetting('adsEnabled', 'true')
+          break
+        case NewTab.RewardsResult.LEDGER_OK:
+          state.rewardsState.walletCreateFailed = true
+          state.rewardsState.walletCreating = false
+          state.rewardsState.walletCreated = false
+          state.rewardsState.walletCorrupted = false
+          break
+      }
+      break
+    }
+
+    case types.ON_REWARDS_SETTING_SAVE:
+      state = { ...state }
```

Comment:
We should only copy state to a new object if we're actually manipulating it. In this `case`, there's a conditional which means sometimes the state won't actually be changed but a re-render will still occur as `state` will be a new object.

Reply:
fixed